### PR TITLE
Thread Safe Mutable Dictionary

### PR DESCRIPTION
**Pull request by rafael-chavez**
@bhariharan @khawkins @sgoldberg-sfdc

### CODE REVIEW

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFOSafeMutableDictionary.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import <Foundation/Foundation.h>
+
+@interface SFOSafeMutableDictionary : NSObject
```
Comment (sgoldberg-sfdc) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFOSafeMutableDictionary.h]: SDK prefix is just SF I believe.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFOSafeMutableDictionary.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
+/*
+ Copyright (c) 2012-present, salesforce.com, inc. All rights reserved.
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFOSafeMutableDictionary.h]: `2017-present`.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFOSafeMutableDictionary.h (hunk 1 above)
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFOSafeMutableDictionary.h]: Actually, we'd like `SFSDK` prefix for anything new going forward.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFOSafeMutableDictionary.m [hunk 3]
```diff
@@ -0,0 +1,117 @@
+/*
+ Copyright (c) 2012-present, salesforce.com, inc. All rights reserved.
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFOSafeMutableDictionary.m]: `2017-present`.

File: libs/SalesforceSDKCore/SalesforceSDKCore/SFOSafeMutableDictionaryTests.m [hunk 4]
```diff
@@ -0,0 +1,100 @@
+/*
+ Copyright (c) 2012-present, salesforce.com, inc. All rights reserved.
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/SFOSafeMutableDictionaryTests.m]: `2017-present`.

File: libs/SalesforceSDKCore/SalesforceSDKCore/SFOSafeMutableDictionaryTests.m [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import <XCTest/XCTest.h>
+#import "SFOSafeMutableDictionary.h"
+
+@interface SFOSafeMutableDictionaryTests : XCTestCase
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/SFOSafeMutableDictionaryTests.m]: Move this file under the `SalesforceSDKCoreTests` folder.

File: libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -42,6 +42,9 @@
 		05460B721DD409B8004E3320 /* SFSwiftDetectUtil.m in Sources */ = {isa = PBXBuildFile; fileRef = 05460B701DD409B8004E3320 /* SFSwiftDetectUtil.m */; };
 		1098078C1CB8713A002AF771 /* UIColor+SFColors.m in Sources */ = {isa = PBXBuildFile; fileRef = E1DDC0FD1CAA2A8B002F51DD /* UIColor+SFColors.m */; };
 		444B95D01E83251900908C61 /* UIColor+SFColorsTests.m in Sources */ = {isa = PBXBuildFile; fileRef = 444B95CF1E83251900908C61 /* UIColor+SFColorsTests.m */; };
+		44EFFA311EE8C63600B30267 /* SFOSafeMutableDictionary.h in Headers */ = {isa = PBXBuildFile; fileRef = 44EFFA2F1EE8C63600B30267 /* SFOSafeMutableDictionary.h */; };
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj]: Make sure the new files are a part of both the static and dynamic targets (`SalesforceSDKCore` and `SalesforceSDKCore-Static`).

**Comment by rafael-chavez** [comment]
updated @bhariharan !

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFSDKSafeMutableDictionary.m [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+@end
+
+@implementation SFSDKSafeMutableDictionary
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        self.backingDictionary = [NSMutableDictionary new];
+        self.queue = dispatch_queue_create([NSString stringWithFormat:@"com.salesforce.chatter.readWriteQueue%d", arc4random_uniform(UINT32_MAX)].UTF8String, DISPATCH_QUEUE_CONCURRENT);
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFSDKSafeMutableDictionary.m]: Minor: this should be `%u`.

Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFSDKSafeMutableDictionary.m]: Replace `com.salesforce.chatter.readWriteQueue` with something more generic like `com.salesforce.mobilesdk.readWriteQueue`.

File: libs/SalesforceSDKCore/SalesforceSDKCore/SFSDKSafeMutableDictionaryTests.m [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import <XCTest/XCTest.h>
+#import "SFSDKSafeMutableDictionary.h"
+
+@interface SFSDKSafeMutableDictionaryTests : XCTestCase
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/SFSDKSafeMutableDictionaryTests.m]: Remove this file. Looks like there are 2 copies of this file.

File: libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -972,6 +977,9 @@
 		05460B6F1DD409B8004E3320 /* SFSwiftDetectUtil.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = SFSwiftDetectUtil.h; sourceTree = "<group>"; };
 		05460B701DD409B8004E3320 /* SFSwiftDetectUtil.m */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.objc; path = SFSwiftDetectUtil.m; sourceTree = "<group>"; };
 		444B95CF1E83251900908C61 /* UIColor+SFColorsTests.m */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.objc; name = "UIColor+SFColorsTests.m"; path = "SalesforceSDKCoreTests/UIColor+SFColorsTests.m"; sourceTree = SOURCE_ROOT; };
+		44D352381EE9AAD000D0E677 /* SFSDKSafeMutableDictionaryTests.m */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.objc; path = SFSDKSafeMutableDictionaryTests.m; sourceTree = "<group>"; };
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj]: Couple of things:

1. This file got dropped into the root of SalesforceSDKCore.
2. The original file, while not present in the project, is still physically present at libs/SalesforceSDKCore/SalesforceSDKCoreTests/SFSDKSafeMutableDictionaryTests.m.

There's some cleanup that needs to happen there.

**Pull request merged**